src/backend/parser/parse_clause.c

```diff
@@ -151,6 +152,13 @@ transformFromClause(ParseState *pstate, List *frmList)
 	 * but those should have been that way already.
 	 */
 	setNamespaceLateralState(pstate->p_namespace, false, true);
+
+	/* 
+	 * Save the namespace -- tsql needs the leftmost select's namespace to
+	 * resolve some ORDER BY clauses used with set operations (i.e. UNION) 
+	 */
+	if (sql_dialect == SQL_DIALECT_TSQL && set_op_ns_stack && !set_op_ns_stack->namespace)
+		set_op_ns_stack->namespace = pstate->p_namespace;
```

Comment:
we should add a hook named post_transfrom_clause_hook, and move the impl into babel